### PR TITLE
3 Create the option to use own prices / self provided prices

### PR DESCRIPTION
**Pull request by ArdJonker**
Completed the functionality to:
+ Read prices/emissions from a third party integration (e.g. Amber Electric)
+ Process these prices/emissions
+ Upload them to FlexMeasures.
+ Request a new schedule if prices/emissions have changed.

### CODE REVIEW

File: appdaemon/apps/v2g-liberty/manage_own_price_data.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        consumption_prices = []
+        emissions = []
+
+        state = await self.get_state(self.HA_CPF_SENSOR_ID, attribute="all")
+        collection_cpf = state["attributes"][self.COLLECTION_NAME]
+        for item in collection_cpf:
+            consumption_prices.append(round(float(item[self.PRICE_LABEL]) * self.KWH_MWH_FACTOR, 2))
+            # Emissions are same for consumption- and production prices so only done here.
+            # Amber has % renewables
+            # we emissions % = 100 - renewables %.
```
Comment (Flix6x) [inline appdaemon/apps/v2g-liberty/manage_own_price_data.py:125]: Rationale: FlexMeasures runs a cost **minimization**.

File: appdaemon/apps/v2g-liberty/manage_own_price_data.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+class ManageOwnPriceData(hass.Hass):
+    """
+    App reads data from the entities that the integration of the power-company
+    writes.
+    It is expected that there are:
+    Forecast entities for consumption- and production prices per kWh for fixed intervals in
+    a longer period ahead (e.g 12 or 24 hrs). These also include an indication of the
+    emissions/kWh.
+    There are entities for a fixed consumption- and production price for the next interval.
```
Comment (Flix6x) [inline appdaemon/apps/v2g-liberty/manage_own_price_data.py:20]: An (abbreviated) example of such an entity would help this docstring.

**Review by Flix6x — CHANGES_REQUESTED** [review]
Mostly requesting extra documentation. A potential follow-up would then be to discuss current limitations of the input entity and potential for generalization.

File: appdaemon/apps/v2g-liberty/get_fm_data.py [hunk 3]
```diff
@@ -423,12 +423,21 @@ def get_emission_intensities(self, *args, **kwargs):
...
-            emission_value = int(round(float(emission_value) / 10, 0))
+            emission_value = int(float(emission_value))
             period_start = datetime.fromtimestamp(emission['event_start'] / 1000).isoformat()
             # ToDO: only make and add data_point if less then 5 hours old this keeps the graph clean.
             data_point = {'time': period_start, 'emission': emission_value}
             emission_points.append(data_point)
 
+        if c.ELECTRICITY_PROVIDER != "self_provided":
+            # Prepare for showing in the chart whereby values are mapped to the 0 - 100% SoC axes.
+            # The max value in the list should match 95%. Expected are only positive values.
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply appdaemon/apps/v2g-liberty/get_fm_data.py:434]: Indeed, will explain in comment

File: README.md [hunk 4]
```diff
@@ -283,6 +283,30 @@ fm_own_price_consumption_sensor_id: cc
 fm_own_emissions_sensor_id: ee
 fm_own_context_display_name: "Own Prices and Emissions"
 
+# The "own price configuration" is mend to be "on the way to generic" but currently
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply README.md:286]: Will correct this

File: appdaemon/apps/v2g-liberty/manage_own_price_data.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    """
+
+    FM_ENTITY_PRICE_PRODUCTION: str
+    FM_ENTITY_PRICE_CONSUMPTION: str
+    FM_ENTITY_EMISSIONS: str
+
+    HA_CPF_SENSOR_ID: str
+    HA_PPF_SENSOR_ID: str
+    HA_CPX_SENSOR_ID: str
+    HA_PPX_SENSOR_ID: str
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply appdaemon/apps/v2g-liberty/manage_own_price_data.py:30]: Will add comments

File: appdaemon/apps/v2g-liberty/manage_own_price_data.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+    COLLECTION_NAME: str
+    PRICE_LABEL: str
+    EMISSION_LABEL: str
+    UOM_LABEL: str
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply appdaemon/apps/v2g-liberty/manage_own_price_data.py:35]: These all need comments, UOM: Unit of Measure.

File: appdaemon/apps/v2g-liberty/manage_own_price_data.py (hunk 1 above)
Comment (ArdJonker) [reply appdaemon/apps/v2g-liberty/manage_own_price_data.py:125]: will add

File: appdaemon/apps/v2g-liberty/manage_own_price_data.py (hunk 2 above)
Comment (ArdJonker) [reply appdaemon/apps/v2g-liberty/manage_own_price_data.py:20]: will add

File: appdaemon/apps/v2g-liberty/get_fm_data.py [hunk 7]
```diff
@@ -423,12 +423,21 @@ def get_emission_intensities(self, *args, **kwargs):
...
+            # TODO: Improve
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply appdaemon/apps/v2g-liberty/get_fm_data.py]: Removed

**Pull request merged**